src/RTCPeerConnection.ts

```diff
@@ -553,7 +553,9 @@ export default class RTCPeerConnection extends defineCustomEventTarget(...PEER_C
             this.dispatchEvent(new RTCTrackEvent('track', eventData));
 
             // Dispatch an unmute event for the track.
-            track._setMutedInternal(false);
+            if (track instanceof MediaStreamTrack) {
```

Comment:
When would this not be an actual track?

Reply:
When I triger the peerConnectionOnTrack event, 
the line 489 "if (oldTransceiver) " always true, and line 500 "track = transceiver._receiver._track;" will cause track haven't include function "_setMutedInternal".

Actually, I am not sure the "if (oldTransceiver) " for what.  :sweat_smile:

Reply:
OnTrack should never fire for an undefined track. What exact scenario allows you to trigger that behavior?

Reply:
A peer createOffer --> B peer receive Offer --> B peer setRemoteDescription(Offer), then the "peerConnectionOnTrack" event be triggered.

Reply:
We'll sure, but how does the SDP look like to have the track be undefined?

Reply:
When ready state is ended, the track._setMutedInternal will be undefined.
##### track log:
```
{
    "readyState": "Ended",
    "id": "d549ef93-835a-4ee7-88e0-34bb7aa523da",
    "peerConnectionId": 0,
    "enabled": true,
    "kind": "video",
    "remote": true
}
```
the track._setMutedInternal is normal going if ready state is live 
##### track log:
```
{
    "label": "",
    "_constraints": {},
    "_enabled": true,
    "_settings": {},
    "_muted": false,
    "_peerConnectionId": 0,
    "id": "d6ce2a2b-258f-46e3-aefd-162f1d2d58f3",
    "kind": "audio",
    "remote": true,
    "readyState": "live"
}
```

Reply:
Yes, that's the other part of the fix. I'm asking about OnTrack with an undefined track. How do you reproduce that?

Reply:
How to reproduce the track undefined:
1. I complete the offer and answer to A/B peers.
2. and complete to add candidate to A/B peers.
3. Fire A peer to get the tracks of localstream and then add these tracks into peer connection.
4. After step 3, the "peerConnectionOnTrack" event will be triggered ,and I log the even as follow:
``` json
{
    "streams": [],
    "transceiverOrder": 2,
    "pcId": 0,
    "receiver": {
        "track": {
            "readyState": "Ended",
            "id": "b5399369-7577-4e62-9da1-76d81dc74236",
            "peerConnectionId": 0,
            "enabled": true,
            "kind": "audio",
            "remote": true
        },
        "rtpParameters": {
            "codecs": [],
            "rtcp": {
                "reducedSize": false,
                "cname": ""
            },
            "headerExtensions": [],
            "encodings": [
                {
                    "active": true
                }
            ],
            "transactionId": ""
        },
        "peerConnectionId": 0,
        "id": "b5399369-7577-4e62-9da1-76d81dc74236"
    },
    "transceiver": {
        "receiver": {
            "id": "b5399369-7577-4e62-9da1-76d81dc74236",
            "peerConnectionId": 0,
            "rtpParameters": {
                "transactionId": "",
                "encodings": [
                    {
                        "active": true
                    }
                ],
                "headerExtensions": [],
                "rtcp": {
                    "cname": "",
                    "reducedSize": false
                },
                "codecs": []
            },
            "track": {
                "remote": true,
                "kind": "audio",
                "enabled": true,
                "peerConnectionId": 0,
                "id": "b5399369-7577-4e62-9da1-76d81dc74236",
                "readyState": "Ended"
            }
        },
        "peerConnectionId": 0,
        "isStopped": false,
        "mid": "0",
        "direction": "recvonly",
        "id": "be53991e-357d-4bbd-a2e6-96cf92bcfd00",
        "sender": {
            "id": "be53991e-357d-4bbd-a2e6-96cf92bcfd00",
            "peerConnectionId": 0,
            "rtpParameters": {
                "transactionId": "ebcfad47-cb0d-4bbc-a84a-8bb291630dc6",
                "encodings": [],
                "headerExtensions": [],
                "rtcp": {
                    "cname": "",
                    "reducedSize": false
                },
                "codecs": []
            }
        }
    }
}
```
this event will cause "track._setMutedInternal(false);" to get 
``` 
Uncaught Error 
undefined is not a function
```
I hope this log is helpful for you.

Reply:
Did you already have a transceiver, or did this last OA create a new one? If it's the latter, I see a `track` definition, so it would certainly create a matching `MediaStreamTrack` object.

If it's the former, can you please log the transceivers (including receiovers) at the beginning of the ontrack function? (not on your code, add some logs to this package's ontrack handler.

Reply:
It's the former.
(A peer offer to B peer, and B answer to A)

When I call the setRemoteDescription function, the package will create a new Transceiver as follows:
B peer's RTCSessionDescription
```json
 {
    "sdp": "v=0\r\no=- 1754350033340842201 2 IN IP4 127.0.0.1\r\ns=-\r\nt=0 0\r\na=group:BUNDLE 0 1\r\na=extmap-allow-mixed\r\na=msid-semantic: WMS\r\nm=audio 9 UDP/TLS/RTP/SAVPF 111 63 103 104 9 102 0 8 106 105 13 110 112 113 126\r\nc=IN IP4 0.0.0.0\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=ice-ufrag:Ak0x\r\na=ice-pwd:6OwDNpXlgBqKqxLwXvaMZjR8\r\na=ice-options:trickle renomination\r\na=fingerprint:sha-256 40:C8:0D:F2:C0:0B:F4:CC:46:95:C6:D7:76:58:CA:B9:1A:C3:01:F0:88:74:D2:4F:88:D6:66:95:54:00:B8:6B\r\na=setup:actpass\r\na=mid:0\r\na=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level\r\na=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time\r\na=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01\r\na=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid\r\na=sendrecv\r\na=msid:- 3ef233f8-c559-47f8-a987-57d262956966\r\na=rtcp-mux\r\na=rtpmap:111 opus/48000/2\r\na=rtcp-fb:111 transport-cc\r\na=fmtp:111 minptime=10;useinbandfec=1\r\na=rtpmap:63 red/48000/2\r\na=fmtp:63 111/111\r\na=rtpmap:103 ISAC/16000\r\na=rtpmap:104 ISAC/32000\r\na=rtpmap:9 G722/8000\r\na=rtpmap:102 ILBC/8000\r\na=rtpmap:0 PCMU/8000\r\na=rtpmap:8 PCMA/8000\r\na=rtpmap:106 CN/32000\r\na=rtpmap:105 CN/16000\r\na=rtpmap:13 CN/8000\r\na=rtpmap:110 telephone-event/48000\r\na=rtpmap:112 telephone-event/32000\r\na=rtpmap:113 telephone-event/16000\r\na=rtpmap:126 telephone-event/8000\r\na=ssrc:3525383695 cname:wVcOw2M7rIoW+k8y\r\na=ssrc:3525383695 msid:- 3ef233f8-c559-47f8-a987-57d262956966\r\nm=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 35 36 125 124 127\r\nc=IN IP4 0.0.0.0\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=ice-ufrag:Ak0x\r\na=ice-pwd:6OwDNpXlgBqKqxLwXvaMZjR8\r\na=ice-options:trickle renomination\r\na=fingerprint:sha-256 40:C8:0D:F2:C0:0B:F4:CC:46:95:C6:D7:76:58:CA:B9:1A:C3:01:F0:88:74:D2:4F:88:D6:66:95:54:00:B8:6B\r\na=setup:actpass\r\na=mid:1\r\na=extmap:14 urn:ietf:params:rtp-hdrext:toffset\r\na=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time\r\na=extmap:13 urn:3gpp:video-orientation\r\na=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01\r\na=extmap:5 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay\r\na=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type\r\na=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing\r\na=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/color-space\r\na=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid\r\na=extmap:10 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id\r\na=extmap:11 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id\r\na=sendrecv\r\na=msid:- 4f7e5a44-57d2-4567-aa2d-5a5202980b65\r\na=rtcp-mux\r\na=rtcp-rsize\r\na=rtpmap:96 VP8/90000\r\na=rtcp-fb:96 goog-remb\r\na=rtcp-fb:96 transport-cc\r\na=rtcp-fb:96 ccm fir\r\na=rtcp-fb:96 nack\r\na=rtcp-fb:96 nack pli\r\na=rtpmap:97 rtx/90000\r\na=fmtp:97 apt=96\r\na=rtpmap:98 VP9/90000\r\na=rtcp-fb:98 goog-remb\r\na=rtcp-fb:98 transport-cc\r\na=rtcp-fb:98 ccm fir\r\na=rtcp-fb:98 nack\r\na=rtcp-fb:98 nack pli\r\na=rtpmap:99 rtx/90000\r\na=fmtp:99 apt=98\r\na=rtpmap:35 AV1/90000\r\na=rtcp-fb:35 goog-remb\r\na=rtcp-fb:35 transport-cc\r\na=rtcp-fb:35 ccm fir\r\na=rtcp-fb:35 nack\r\na=rtcp-fb:35 nack pli\r\na=rtpmap:36 rtx/90000\r\na=fmtp:36 apt=35\r\na=rtpmap:125 red/90000\r\na=rtpmap:124 rtx/90000\r\na=fmtp:124 apt=125\r\na=rtpmap:127 ulpfec/90000\r\na=ssrc-group:FID 3019939996 1053936332\r\na=ssrc:3019939996 cname:wVcOw2M7rIoW+k8y\r\na=ssrc:3019939996 msid:- 4f7e5a44-57d2-4567-aa2d-5a5202980b65\r\na=ssrc:1053936332 cname:wVcOw2M7rIoW+k8y\r\na=ssrc:1053936332 msid:- 4f7e5a44-57d2-4567-aa2d-5a5202980b65\r\n",
    "type": "offer"
}
```
B peer get the new Transceivers, audio
```json
{
    "_mid": "0",
    "_peerConnectionId": 0,
    "_id": "fff0d798-6c0a-45f0-9221-6458e9d301e0",
    "_direction": "recvonly",
    "_stopped": false,
    "_sender": {
        "_track": null,
        "_peerConnectionId": 0,
        "_id": "fff0d798-6c0a-45f0-9221-6458e9d301e0",
        "_rtpParameters": {
            "codecs": [],
            "headerExtensions": [],
            "rtcp": {
                "cname": "",
                "reducedSize": false
            },
            "transactionId": "dc2242f7-d82b-490a-b5e7-566b672f7bec",
            "encodings": []
        }
    },
    "_receiver": {
        "_track": {
            "readyState": "Ended",
            "id": "3ef233f8-c559-47f8-a987-57d262956966",
            "peerConnectionId": 0,
            "enabled": true,
            "kind": "audio",
            "remote": true
        },
        "_id": "3ef233f8-c559-47f8-a987-57d262956966",
        "_peerConnectionId": 0,
        "_rtpParameters": {
            "codecs": [],
            "headerExtensions": [],
            "rtcp": {
                "cname": "",
                "reducedSize": false
            }
        }
    }
}
```
and a video
```json
 {
    "_mid": "1",
    "_peerConnectionId": 0,
    "_id": "1053fc0b-4cca-46b7-8a88-42d4ac9bd9ba",
    "_direction": "recvonly",
    "_stopped": false,
    "_sender": {
        "_track": null,
        "_peerConnectionId": 0,
        "_id": "1053fc0b-4cca-46b7-8a88-42d4ac9bd9ba",
        "_rtpParameters": {
            "codecs": [],
            "headerExtensions": [],
            "rtcp": {
                "cname": "",
                "reducedSize": false
            },
            "transactionId": "be9efac6-6dcb-4f54-85bd-d666ee2449f7",
            "encodings": []
        }
    },
    "_receiver": {
        "_track": {
            "readyState": "Ended",
            "id": "4f7e5a44-57d2-4567-aa2d-5a5202980b65",
            "peerConnectionId": 0,
            "enabled": true,
            "kind": "video",
            "remote": true
        },
        "_id": "4f7e5a44-57d2-4567-aa2d-5a5202980b65",
        "_peerConnectionId": 0,
        "_rtpParameters": {
            "codecs": [],
            "headerExtensions": [],
            "rtcp": {
                "cname": "",
                "reducedSize": true
            }
        }
    }
}
```
To cause the oldTransceiver always is true in peerConnectionOnTrack event listener, _transceivers log as follows:
``` javascript
[
    {
        "order": 0,
        "transceiver": {
            "_mid": "0",
            "_peerConnectionId": 0,
            "_id": "fff0d798-6c0a-45f0-9221-6458e9d301e0",
            "_direction": "recvonly",
            "_stopped": false,
            "_sender": {
                "_track": null,
                "_peerConnectionId": 0,
                "_id": "fff0d798-6c0a-45f0-9221-6458e9d301e0",
                "_rtpParameters": {
                    "codecs": [],
                    "headerExtensions": [],
                    "rtcp": {
                        "cname": "",
                        "reducedSize": false
                    },
                    "transactionId": "dc2242f7-d82b-490a-b5e7-566b672f7bec",
                    "encodings": []
                }
            },
            "_receiver": {
                "_track": {
                    "readyState": "Ended",
                    "id": "3ef233f8-c559-47f8-a987-57d262956966",
                    "peerConnectionId": 0,
                    "enabled": true,
                    "kind": "audio",
                    "remote": true
                },
                "_id": "3ef233f8-c559-47f8-a987-57d262956966",
                "_peerConnectionId": 0,
                "_rtpParameters": {
                    "codecs": [],
                    "headerExtensions": [],
                    "rtcp": {
                        "cname": "",
                        "reducedSize": false
                    }
                }
            }
        }
    },
    {
        "order": 1,
        "transceiver": {
            "_mid": "1",
            "_peerConnectionId": 0,
            "_id": "1053fc0b-4cca-46b7-8a88-42d4ac9bd9ba",
            "_direction": "recvonly",
            "_stopped": false,
            "_sender": {
                "_track": null,
                "_peerConnectionId": 0,
                "_id": "1053fc0b-4cca-46b7-8a88-42d4ac9bd9ba",
                "_rtpParameters": {
                    "codecs": [],
                    "headerExtensions": [],
                    "rtcp": {
                        "cname": "",
                        "reducedSize": false
                    },
                    "transactionId": "be9efac6-6dcb-4f54-85bd-d666ee2449f7",
                    "encodings": []
                }
            },
            "_receiver": {
                "_track": {
                    "readyState": "Ended",
                    "id": "4f7e5a44-57d2-4567-aa2d-5a5202980b65",
                    "peerConnectionId": 0,
                    "enabled": true,
                    "kind": "video",
                    "remote": true
                },
                "_id": "4f7e5a44-57d2-4567-aa2d-5a5202980b65",
                "_peerConnectionId": 0,
                "_rtpParameters": {
                    "codecs": [],
                    "headerExtensions": [],
                    "rtcp": {
                        "cname": "",
                        "reducedSize": true
                    }
                }
            }
        }
    }
]
```
I am not sure, why the 2 Transceivers will cause track._setMutedInternal be undefined.